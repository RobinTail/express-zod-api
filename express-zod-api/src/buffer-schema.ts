import { z } from "zod/v4";

export const bufferSchema = z.custom<Buffer>(
  (subject) => Buffer.isBuffer(subject),
  { error: "Expected Buffer" },
);
