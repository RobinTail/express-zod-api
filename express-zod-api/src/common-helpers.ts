import { Request } from "express";
import * as R from "ramda";
import { z } from "zod";
import { CommonConfig, InputSource, InputSources } from "./config-type";
import { contentTypes } from "./content-type";
import { OutputValidationError } from "./errors";
import { metaSymbol } from "./metadata";
import { AuxMethod, Method } from "./method";

/** @desc this type does not allow props assignment, but it works for reading them when merged with another interface */
export type EmptyObject = Record<string, never>;
export type EmptySchema = z.ZodObject<EmptyObject, "strip">;
export type FlatObject = Record<string, unknown>;

/** @link https://stackoverflow.com/a/65492934 */
type NoNever<T, F> = [T] extends [never] ? F : T;

/**
 * @desc Using module augmentation approach you can specify tags as the keys of this interface
 * @example declare module "express-zod-api" { interface TagOverrides { users: unknown } }
 * @link https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation
 * */
// eslint-disable-next-line @typescript-eslint/no-empty-object-type -- augmentation
export interface TagOverrides {}
export type Tag = NoNever<keyof TagOverrides, string>;

/** @see https://expressjs.com/en/guide/routing.html */
export const routePathParamsRegex = /:([A-Za-z0-9_]+)/g;
export const getRoutePathParams = (path: string): string[] =>
  path.match(routePathParamsRegex)?.map((param) => param.slice(1)) || [];

const areFilesAvailable = (request: Request): boolean => {
  const contentType = request.header("content-type") || "";
  const isUpload = contentType.toLowerCase().startsWith(contentTypes.upload);
  return "files" in request && isUpload;
};

export const defaultInputSources: InputSources = {
  get: ["query", "params"],
  post: ["body", "params", "files"],
  put: ["body", "params"],
  patch: ["body", "params"],
  delete: ["query", "params"],
};
const fallbackInputSource: InputSource[] = ["body", "query", "params"];

export const getActualMethod = (request: Request) =>
  request.method.toLowerCase() as Method | AuxMethod;

export const getInput = (
  req: Request,
  userDefined: CommonConfig["inputSources"] = {},
): FlatObject => {
  const method = getActualMethod(req);
  if (method === "options") return {};
  return (
    userDefined[method] ||
    defaultInputSources[method] ||
    fallbackInputSource
  )
    .filter((src) => (src === "files" ? areFilesAvailable(req) : true))
    .reduce<FlatObject>((agg, src) => Object.assign(agg, req[src]), {});
};

export const ensureError = (subject: unknown): Error =>
  subject instanceof Error ? subject : new Error(String(subject));

export const getMessageFromError = (error: Error): string => {
  if (error instanceof z.ZodError) {
    return error.issues
      .map(({ path, message }) =>
        (path.length ? [path.join("/")] : []).concat(message).join(": "),
      )
      .join("; ");
  }
  if (error instanceof OutputValidationError) {
    const hasFirstField = error.cause.issues[0]?.path.length > 0;
    return `output${hasFirstField ? "/" : ": "}${error.message}`;
  }
  return error.message;
};

/** Takes the original unvalidated examples from the properties of ZodObject schema shape */
export const pullExampleProps = <T extends z.SomeZodObject>(subject: T) =>
  Object.entries(subject.shape).reduce<Partial<z.input<T>>[]>(
    (acc, [key, schema]) => {
      const { _def } = schema as z.ZodType;
      return combinations(
        acc,
        (_def[metaSymbol]?.examples || []).map(R.objOf(key)),
        ([left, right]) => ({ ...left, ...right }),
      );
    },
    [],
  );

export const getExamples = <
  T extends z.ZodType,
  V extends "original" | "parsed" | undefined,
>({
  schema,
  variant = "original",
  validate = variant === "parsed",
  pullProps = false,
}: {
  schema: T;
  /**
   * @desc examples variant: original or parsed
   * @example "parsed" — for the case when possible schema transformations should be applied
   * @default "original"
   * @override validate: variant "parsed" activates validation as well
   * */
  variant?: V;
  /**
   * @desc filters out the examples that do not match the schema
   * @default variant === "parsed"
   * */
  validate?: boolean;
  /**
   * @desc should pull examples from properties — applicable to ZodObject only
   * @default false
   * */
  pullProps?: boolean;
}): ReadonlyArray<V extends "parsed" ? z.output<T> : z.input<T>> => {
  let examples = schema._def[metaSymbol]?.examples || [];
  if (!examples.length && pullProps && schema instanceof z.ZodObject)
    examples = pullExampleProps(schema);
  if (!validate && variant === "original") return examples;
  const result: Array<z.input<T> | z.output<T>> = [];
  for (const example of examples) {
    const parsedExample = schema.safeParse(example);
    if (parsedExample.success)
      result.push(variant === "parsed" ? parsedExample.data : example);
  }
  return result;
};

export const combinations = <T>(
  a: T[],
  b: T[],
  merge: (pair: [T, T]) => T,
): T[] => (a.length && b.length ? R.xprod(a, b).map(merge) : a.concat(b));

/**
 * @desc isNullable() and isOptional() validate the schema's input
 * @desc They always return true in case of coercion, which should be taken into account when depicting response
 */
export const hasCoercion = (schema: z.ZodTypeAny): boolean =>
  "coerce" in schema._def && typeof schema._def.coerce === "boolean"
    ? schema._def.coerce
    : false;

export const ucFirst = (subject: string) =>
  subject.charAt(0).toUpperCase() + subject.slice(1).toLowerCase();

export const makeCleanId = (...args: string[]) => {
  const byAlpha = R.chain((entry) => entry.split(/[^A-Z0-9]/gi), args);
  const byWord = R.chain(
    (entry) =>
      entry.replaceAll(/[A-Z]+/g, (beginning) => `/${beginning}`).split("/"),
    byAlpha,
  );
  return byWord.map(ucFirst).join("");
};

export const getTransformedType = R.tryCatch(
  <T>(schema: z.ZodEffects<z.ZodTypeAny, unknown, T>, sample: T) =>
    typeof schema.parse(sample),
  R.always(undefined),
);

/** @desc can still be an array, use Array.isArray() or rather R.type() to exclude that case */
export const isObject = (subject: unknown) =>
  typeof subject === "object" && subject !== null;

export const isProduction = R.memoizeWith(
  () => process.env.TSUP_STATIC as string, // eslint-disable-line no-restricted-syntax -- substituted by TSUP
  () => process.env.NODE_ENV === "production", // eslint-disable-line no-restricted-syntax -- memoized
);
