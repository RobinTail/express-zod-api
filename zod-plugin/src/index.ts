import "./runtime.ts"; // side effects here
export { pack, unpack } from "./packer.ts";
export { getBrand } from "./brand.ts";
