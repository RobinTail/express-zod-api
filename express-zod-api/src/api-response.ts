import { z } from "zod";
import { isSchema } from "./common-helpers";
import { contentTypes } from "./content-type";

export const defaultStatusCodes = {
  positive: 200,
  negative: 400,
} satisfies Record<string, number>;

export type ResponseVariant = keyof typeof defaultStatusCodes;
export const responseVariants = Object.keys(
  defaultStatusCodes,
) as ResponseVariant[];

/** @public this is the user facing configuration */
export class ApiResponse<S extends z.ZodType> {
  readonly #schema: S;
  readonly #statusCodes?: [number, ...number[]];
  readonly #mimeTypes?: [string, ...string[]] | null;

  public constructor(
    subject:
      | S // shorthand for { schema: S }, default status codes and MIME types applied
      | {
          schema: S;
          /** @default 200 for a positive and 400 for a negative response */
          statusCode?: number | [number, ...number[]];
          /**
           * @example null is for no content, such as 204 and 302
           * @default "application/json"
           * */
          mimeType?: string | [string, ...string[]] | null;
        },
  ) {
    const { schema, statusCode, mimeType } = isSchema(subject)
      ? { schema: subject }
      : subject;
    this.#schema = schema;
    this.#statusCodes =
      typeof statusCode === "number" ? [statusCode] : statusCode;
    this.#mimeTypes = typeof mimeType === "string" ? [mimeType] : mimeType;
  }

  public normalize(fbStatus: number) {
    return {
      schema: this.#schema,
      statusCodes: this.#statusCodes ?? [fbStatus],
      mimeTypes:
        this.#mimeTypes === undefined ? [contentTypes.json] : this.#mimeTypes,
    };
  }
}

export type NormalizedResponse = ReturnType<
  ApiResponse<z.ZodType>["normalize"]
>;
