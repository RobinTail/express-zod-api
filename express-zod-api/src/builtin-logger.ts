import ansis from "ansis";
import { inspect } from "node:util";
import { performance } from "node:perf_hooks";
import { FlatObject, isProduction } from "./common-helpers";
import {
  AbstractLogger,
  formatDuration,
  isHidden,
  Severity,
  styles,
} from "./logger-helpers";

interface Context extends FlatObject {
  requestId?: string;
}

export interface BuiltinLoggerConfig {
  /**
   * @desc The minimal severity to log or "silent" to disable logging
   * @example "debug" also enables pretty output for inspected entities
   * */
  level: "silent" | "warn" | "info" | "debug";
  /** @desc Enables colors on printed severity and inspected entities */
  color: boolean;
  /**
   * @desc Printing logs does not delay execution
   * @see https://expressjs.com/en/advanced/best-practice-performance.html#do-logging-correctly
   * */
  async: boolean;
  /**
   * @desc Control how deeply entities should be inspected
   * @example null
   * @example Infinity
   * */
  depth: number | null;
  /**
   * @desc Context: the metadata applicable for each logged entry, used by .child() method
   * @see childLoggerProvider
   * */
  ctx: Context;
}

interface ProfilerOptions {
  message: string;
  /** @default "debug" */
  severity?: Severity | ((ms: number) => Severity);
  /** @default formatDuration - adaptive units and limited fraction */
  formatter?: (ms: number) => string | number;
}

/** @desc Built-in console logger with optional colorful inspections */
export class BuiltinLogger implements AbstractLogger {
  protected readonly config: BuiltinLoggerConfig;
  protected readonly stack: string[] = [];
  protected postponed?: NodeJS.Immediate;

  /** @example new BuiltinLogger({ level: "debug", color: true, depth: 4 }) */
  public constructor({
    color = ansis.isSupported(),
    level = isProduction() ? "warn" : "debug",
    async = isProduction(),
    depth = 2,
    ctx = {},
  }: Partial<BuiltinLoggerConfig> = {}) {
    this.config = { color, level, async, depth, ctx };
  }

  protected format(subject: unknown) {
    const { depth, color: colors, level } = this.config;
    return inspect(subject, {
      depth,
      colors,
      breakLength: level === "debug" ? 80 : Infinity,
      compact: level === "debug" ? 3 : true,
    });
  }

  protected print(method: Severity, message: string, meta?: unknown) {
    const {
      level,
      ctx: { requestId, ...ctx },
      color: hasColor,
      async: isAsync,
    } = this.config;
    if (level === "silent" || isHidden(method, level)) return;
    const output: string[] = [new Date().toISOString()];
    if (requestId) output.push(hasColor ? styles.ctx(requestId) : requestId);
    output.push(
      hasColor ? `${styles[method](method)}:` : `${method}:`,
      message,
    );
    if (meta !== undefined) output.push(this.format(meta));
    if (Object.keys(ctx).length > 0) output.push(this.format(ctx));
    (isAsync ? this.postpone.bind(this) : console.log)(output.join(" "));
  }

  protected postpone(line: string) {
    this.stack.push(line);
    clearImmediate(this.postponed);
    this.postponed = setImmediate(this.purge.bind(this)).unref(); // do not block process.exit()
  }

  /** @internal */
  public purge() {
    while (this.stack.length) console.log(this.stack.shift());
  }

  /** @internal */
  public dispose(cb: () => void) {
    this.config.level = "silent"; // no more logs
    clearImmediate(this.postponed);
    this.postponed = setImmediate(() => {
      this.purge();
      cb();
    });
  }

  public debug(message: string, meta?: unknown) {
    this.print("debug", message, meta);
  }

  public info(message: string, meta?: unknown) {
    this.print("info", message, meta);
  }

  public warn(message: string, meta?: unknown) {
    this.print("warn", message, meta);
  }

  public error(message: string, meta?: unknown) {
    this.print("error", message, meta);
  }

  public child(ctx: Context) {
    return new BuiltinLogger({ ...this.config, ctx });
  }

  /**
   * @desc The argument used for instance created by .child() method
   * @see ChildLoggerProvider
   * */
  public get ctx() {
    return this.config.ctx;
  }

  /** @desc Measures the duration until you invoke the returned callback */
  public profile(message: string): () => void;
  public profile(options: ProfilerOptions): () => void;
  public profile(subject: string | ProfilerOptions) {
    const start = performance.now();
    return () => {
      const duration = performance.now() - start;
      const {
        message,
        severity = "debug",
        formatter = formatDuration,
      } = typeof subject === "object" ? subject : { message: subject };
      this.print(
        typeof severity === "function" ? severity(duration) : severity,
        message,
        formatter(duration),
      );
    };
  }
}
