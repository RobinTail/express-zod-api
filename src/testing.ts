import { Request, Response } from "express";
import http from "node:http";
import { FlatObject } from "./common-helpers";
import { CommonConfig } from "./config-type";
import { AbstractEndpoint } from "./endpoint";
import { AbstractLogger, ActualLogger } from "./logger-helpers";
import { contentTypes } from "./content-type";
import { loadAlternativePeer } from "./peer-helpers";
import { LocalResponse } from "./server-helpers";

/**
 * @desc Using module augmentation approach you can set the Mock type of your actual testing framework.
 * @example declare module "express-zod-api" { interface MockOverrides extends Mock {} }
 * @link https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation
 * */
export interface MockOverrides {}

/** @desc Compatibility constraints for a function mocking method of a testing framework. */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type MockFunction = (implementation?: (...args: any[]) => any) => MockOverrides;

export const makeRequestMock = <REQ extends FlatObject>({
  fnMethod,
  requestProps,
}: {
  fnMethod: MockFunction;
  requestProps?: REQ;
}) =>
  ({
    method: "GET",
    header: fnMethod(() => contentTypes.json),
    ...requestProps,
  }) as { method: string } & Record<"header", MockOverrides> & REQ;

export const makeResponseMock = <RES extends FlatObject>({
  fnMethod,
  responseProps,
}: {
  fnMethod: MockFunction;
  responseProps?: RES;
}) => {
  const responseMock = {
    writableEnded: false,
    statusCode: 200,
    statusMessage: http.STATUS_CODES[200],
    set: fnMethod(() => responseMock),
    setHeader: fnMethod(() => responseMock),
    header: fnMethod(() => responseMock),
    status: fnMethod((code: number) => {
      responseMock.statusCode = code;
      responseMock.statusMessage = http.STATUS_CODES[code]!;
      return responseMock;
    }),
    json: fnMethod(() => responseMock),
    send: fnMethod(() => responseMock),
    end: fnMethod(() => {
      responseMock.writableEnded = true;
      return responseMock;
    }),
    locals: {},
    ...responseProps,
  } as {
    writableEnded: boolean;
    statusCode: number;
    statusMessage: string;
    locals: LocalResponse["locals"];
  } & Record<
    "set" | "setHeader" | "header" | "status" | "json" | "send" | "end",
    MockOverrides
  > &
    RES;
  return responseMock;
};

export const makeLoggerMock = <LOG extends FlatObject>({
  fnMethod,
  loggerProps,
}: {
  fnMethod: MockFunction;
  loggerProps?: LOG;
}) =>
  ({
    info: fnMethod(),
    warn: fnMethod(),
    error: fnMethod(),
    debug: fnMethod(),
    ...loggerProps,
  }) as Record<keyof AbstractLogger, MockOverrides> & LOG;

interface TestEndpointProps<REQ, RES, LOG> {
  /** @desc The endpoint to test */
  endpoint: AbstractEndpoint;
  /**
   * @desc Additional properties to set on Request mock
   * @default { method: "GET", header: () => "application/json" }
   * */
  requestProps?: REQ;
  /**
   * @desc Additional properties to set on Response mock
   * @default { writableEnded, statusCode, statusMessage, set, setHeader, header, status, json, send, end }
   * */
  responseProps?: RES;
  /**
   * @desc Additional properties to set on config mock
   * @default { cors: false, logger }
   * */
  configProps?: Partial<CommonConfig>;
  /**
   * @desc Additional properties to set on logger mock
   * @default { info, warn, error, debug }
   * */
  loggerProps?: LOG;
  /**
   * @desc Optionally specify the function mocking method of your testing framework
   * @default jest.fn || vi.fn // from vitest
   * @example mock.fn.bind(mock) // from node:test, binding might be necessary
   * */
  fnMethod?: MockFunction;
}

/** @desc Requires either jest (with @types/jest) or vitest or to specify the fnMethod option */
export const testEndpoint = async <
  LOG extends FlatObject,
  REQ extends FlatObject,
  RES extends FlatObject,
>({
  endpoint,
  requestProps,
  responseProps,
  configProps,
  loggerProps,
  fnMethod: userDefined,
}: TestEndpointProps<REQ, RES, LOG>) => {
  const fnMethod =
    userDefined ||
    (
      await loadAlternativePeer<{ fn: MockFunction }>([
        { moduleName: "vitest", moduleExport: "vi" },
        { moduleName: "@jest/globals", moduleExport: "jest" },
      ])
    ).fn;
  const requestMock = makeRequestMock({ fnMethod, requestProps });
  const responseMock = makeResponseMock({ fnMethod, responseProps });
  const loggerMock = makeLoggerMock({ fnMethod, loggerProps });
  const configMock = {
    cors: false,
    logger: loggerMock,
    ...configProps,
  };
  await endpoint.execute({
    request: requestMock as unknown as Request,
    response: responseMock as unknown as Response,
    config: configMock as CommonConfig,
    logger: loggerMock as unknown as ActualLogger,
  });
  return { requestMock, responseMock, loggerMock };
};
