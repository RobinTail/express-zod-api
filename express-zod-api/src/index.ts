import "@express-zod-api/zod-plugin"; // side effects here
export { createConfig } from "./config-type.ts";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory.ts";
export { getMessageFromError } from "./common-helpers.ts";
export { ensureHttpError } from "./result-helpers.ts";
export { BuiltinLogger } from "./builtin-logger.ts";
export { Middleware } from "./middleware.ts";
export {
  ResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler.ts";
export { DependsOnMethod } from "./depends-on-method.ts";
export { ServeStatic } from "./serve-static.ts";
export { createServer, attachRouting } from "./server.ts";
export { Documentation } from "./documentation.ts";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors.ts";
export { testEndpoint, testMiddleware } from "./testing.ts";
export { Integration } from "./integration.ts";
export { EventStreamFactory } from "./sse.ts";

export { ez } from "./proprietary-schemas.ts";

// Convenience types
export type { Depicter } from "./documentation-helpers.ts";
export type { Producer } from "./zts-helpers.ts";

// Interfaces exposed for augmentation
export type { LoggerOverrides } from "./logger-helpers.ts";
export type { TagOverrides } from "./common-helpers.ts";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
import type {} from "qs"; // fixes TS2742 for attachRouting
export type { Routing } from "./routing.ts";
export type { FlatObject } from "./common-helpers.ts";
export type { Method } from "./method.ts";
export type { IOSchema } from "./io-schema.ts";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type.ts";
export type { ApiResponse } from "./api-response.ts";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  HeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security.ts";
