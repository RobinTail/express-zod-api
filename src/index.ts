import "./zod-plugin";

export { createConfig } from "./config-type";
export { AbstractEndpoint } from "./endpoint";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory";
export {
  getExamples,
  getMessageFromError,
  getStatusCodeFromError,
} from "./common-helpers";
export { createLogger } from "./logger";
export { createMiddleware } from "./middleware";
export {
  createResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler";
export { DependsOnMethod } from "./depends-on-method";
export { ServeStatic } from "./serve-static";
export { createServer, attachRouting } from "./server";
export { Documentation } from "./documentation";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors";
export { testEndpoint } from "./testing";
export { Integration } from "./integration";

export { ez } from "./proprietary-schemas";

// Convenience types
export type { Depicter } from "./documentation-helpers";
export type { Producer } from "./zts-helpers";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
export type { MockOverrides } from "./testing";
export type { Routing } from "./routing";
export type { LoggerOverrides } from "./logger";
export type { FlatObject } from "./common-helpers";
export type { Method } from "./method";
export type { IOSchema } from "./io-schema";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type";
export type { MiddlewareDefinition } from "./middleware";
export type { ResultHandlerDefinition } from "./result-handler";
export type { ApiResponse } from "./api-response";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  CustomHeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security";
