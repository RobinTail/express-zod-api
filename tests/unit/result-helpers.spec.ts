import createHttpError from "http-errors";
import { z } from "zod";
import { InputValidationError, OutputValidationError } from "../../src";
import {
  ensureHttpError,
  getPublicErrorMessage,
  logServerError,
} from "../../src/result-helpers";
import { makeLoggerMock, makeRequestMock } from "../../src/testing";

describe("Result helpers", () => {
  describe("logServerError()", () => {
    test("should log server side error", () => {
      const error = createHttpError(501, "test");
      const logger = makeLoggerMock();
      const request = makeRequestMock({ url: "https://example.com" });
      logServerError(error, logger, request, { test: 123 });
      expect(logger._getLogs().error).toEqual([
        [
          "Server side error",
          { error, payload: { test: 123 }, url: "https://example.com" },
        ],
      ]);
    });
  });

  describe("ensureHttpError()", () => {
    test.each([
      new Error("basic"),
      createHttpError(404, "Not really found"),
      new InputValidationError(z.string().safeParse(123).error!),
      new OutputValidationError(z.string().safeParse(123).error!),
    ])("should handle %s", (error) => {
      expect(ensureHttpError(error)).toMatchSnapshot();
    });
  });

  describe.each(["development", "production"])(
    "getPublicErrorMessage() in %s mode",
    (mode) => {
      beforeAll(() => {
        vi.stubEnv("TSUP_STATIC", mode);
        vi.stubEnv("NODE_ENV", mode);
      });
      afterAll(() => vi.unstubAllEnvs());

      test("should return actual message for 400", () => {
        expect(
          getPublicErrorMessage(createHttpError(400, "invalid inputs")),
        ).toBe("invalid inputs");
      });

      test("should comply exposition prop", () => {
        expect(
          getPublicErrorMessage(
            createHttpError(400, "invalid inputs", { expose: false }),
          ),
        ).toBe(mode === "production" ? "Bad Request" : "invalid inputs");
        expect(
          getPublicErrorMessage(
            createHttpError(500, "something particual failed", {
              expose: true,
            }),
          ),
        ).toBe("something particual failed");
      });

      test("should return generalized message for 500", () => {
        expect(
          getPublicErrorMessage(
            createHttpError(500, "something particual failed"),
          ),
        ).toBe(
          mode === "production"
            ? "Internal Server Error"
            : "something particual failed",
        );
      });
    },
  );
});
