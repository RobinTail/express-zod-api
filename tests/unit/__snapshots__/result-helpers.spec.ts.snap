// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Result helpers > ensureHttpError() > should handle Error: basic 1`] = `
InternalServerError({
  "cause": Error({
    "message": "basic",
  }),
  "message": "basic",
})
`;

exports[`Result helpers > ensureHttpError() > should handle InputValidationError: Expected string, received number 1`] = `
BadRequestError({
  "cause": ZodError({
    "message": "[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]",
  }),
  "message": "Expected string, received number",
})
`;

exports[`Result helpers > ensureHttpError() > should handle NotFoundError: Not really found 1`] = `
NotFoundError({
  "message": "Not really found",
})
`;

exports[`Result helpers > ensureHttpError() > should handle OutputValidationError: Expected string, received number 1`] = `
InternalServerError({
  "cause": ZodError({
    "message": "[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [],
    "message": "Expected string, received number"
  }
]",
  }),
  "message": "output: Expected string, received number",
})
`;
